### ISSUE TITLE
Aria role listbox not suitable for Carousels

### ISSUE THREAD

**Issue by bluSCALE4** [express-labs/pure-react-carousel#112]
The argument has been made here:

https://github.com/twbs/bootstrap/issues/22061

**Comment by mrbinky3000** [comment]
There is a discussion going on in the PR.  Unlike bootstrap's carousel, our carousel captures keyboard events so you can use the left and right arrows to navigate.  When you assign event handlers like onkeydown and onclick to a non-interactive HTML element, like a div, then Aria requires that you assign the element a role.  At least that's my understanding of WCAG.

listbox seemed like an option because you can tab to each visible slide in the slider.  However, the audio that is read by a screen reader is a little misleading.  It speaks of "options" to select.  There's no options.  

If anyone has an idea of which role we should use in this situation, please speak up.

**Comment by tim-steele** [comment]
@bluSCALE4 please see the response on PR #113 - I think we don't need to remove the listbox, but I think we could discuss if tablist would be a better option?

**Comment by bluSCALE4** [comment]
@tim-steele I agree, tablist looks like a much better option. Here's what I was reviewing and it makes a lot of sense.

> * Each carousel content container has a role of tabpanel.
> * Each control, typical designed as dots, has a role of tab.
> * The container of the controls/tabs has a role of  tablist.
> * Add aria-labelledby to the tabpanels which point to the id of the associated control/tab.
> * To each control/tab, add aria-controls (which points to the id of the associated tabpanel) and aria-selected (boolean) attributes.

**Comment by bluSCALE4** [comment]
How you guys looked at `role: grid`? Seems to be the catch all of ARIA widget roles. https://www.w3.org/TR/wai-aria-1.1/#grid

**Comment by bluSCALE4** [comment]
Into the rabbits hole we go... found this recent conversation:

https://github.com/w3c/aria-practices/issues/43

**Comment by tim-steele** [comment]
@bluSCALE4 it does look like they are including the spec. I am reviewing their carousel recommendation here https://www.w3.org/TR/wai-aria-practices-1.1/#carousel. I don't think it is final yet.

**Comment by RTG-MColeman** [comment]
It is also important to note that many developers will over use a slider, not as a pure carousel but as a means of selection (ex: a product that has 10 different colors) in this case your list box maybe useful. Also if it is required that a selection be made then maybe even role="radiogroup" to help convey that this is part of a form selection.

There is also possibilities of expanding the useful meaning of the buttons for screen readers (ex: next button should read "Next 3 of 10" or however many slides as visible VS remaining) Same for previous first and last."First 1 of 10" /  "Last 10 of 10"

Also usage of tab seems to be throwing the slider off. In your demo of "Simple Carousel With React Redux" attempt to only use the tab key to move through the slides. The proper order should be Slide 1,2,3 then Next arrow. However as you can see the tab order continues to push the slider all the way through the slides before ever getting to the next button. Even then when the user reaches the next button they do not know they are at the end of the ordered list because the next button is still a valid button. Usage of aria-hidden=true/false and tabindex=0/-1 will keep the hidden slides  out the tab order.

Good luck!

**Comment by hcientist** [comment]
As this [doesn't currently have a clear solution](https://github.com/express-labs/pure-react-carousel/pull/113#issuecomment-454514240), might it make sense for now to default to`role="listbox"` (remaining consistent with the current implementation, but to permit passing in props to override it (e.g. if their use case matches [different guidance](https://w3c.github.io/aria-practices/#basic-carousel-elements))? [Currently](https://github.com/express-labs/pure-react-carousel/blob/5d1c8acfb9c2412281204f03f74b76ad20bcd36c/src/Slider/Slider.jsx#L659) it's not possible to override. Similarly, perhaps the [`aria-roledescription`](https://w3c.github.io/aria/#aria-roledescription) could be passed through to the element?

**Comment by tim-steele** [comment]
@hcientist would you be open to making a Pull Request for making it possible to override? @kylane mentions in Issue #315 the ability to override aria props, so it looks like it something that has support.

**Comment by TwisterMc** [comment]
Where have we landed on this one? Removing `role="listbox"` seems to be a good start or giving us a way to overwrite it.

**Comment by misterbracket** [comment]
We found a very basic workaround to solve these issues for our purposes. 

Without going into too much detail on how the roles work, for us, it was enough to make this work with `role="listbox"`. This change did it for us. 

```js
            {/* Adding role={undefined} and trayProps={{role: 'listbox'}} fixes an issue 
                with accessibility. Basically, the role attribute was on the wrong element.
                The role 'listbox' needs need wrap direct decedents with the role 'option'.
                We hope that this gets addressed by the library soon. */}
            <Slider
                aria-orientation={carouselState.orientation}
                aria-labelledby={carouselId}
                classNameTrayWrap="carousel-tray-wrapper"
                role={undefined}
                trayProps={{role: 'listbox'}}
            >
            ...
            </Slider>
```
